Make 303 redirects do GET like Net Framework

TimothyByrd opened the pull request:
Make 303 redirects do GET like Net Framework

In Net Framework, PUT redirects on a 303 do a GET.
Net 5.0 breaks compatibility with this.
See #28998
This commit causes redirects of a 303 to do a GET
for POST (existing behavior), PUT, DELETE, PATCH and OPTIONS

Fix #28998

Commit:
make PUT redirects act like POST redirects for 30x reponses to conform to net Framework behaviour - issue 28998

Commit:
Make PUT redirects do GET like Net Framework

In Net Framework, PUT redirects do a GET.
Net 5.0 breaks compatibiltiy with this.
See https://github.com/dotnet/runtime/issues/28998
This commit causes PUT redirects to act like POST redirects
for 30x responses to conform to Net Framework behaviour.

Fix #28998

Commit:
Merge branch 'issue/28998' of https://github.com/TimothyByrd/dotnet_runtime into issue/28998

Commit:
Merge branch 'issue/28998' of https://github.com/TimothyByrd/dotnet_runtime into issue/28998

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=49095) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=49095) before we can accept your contribution.<br/><br/>:x: TimothyByrd [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=49095)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=49095) it.</sub>

Commit:
Merge branch 'issue/28998' of https://github.com/TimothyByrd/dotnet_runtime into issue/28998

geoffkizer commented:
I think we need to have consensus on what the right behavior here is in #28998 before we can consider taking this PR.

I'll add my comments there shortly.

Commit:
Make 303 redirects do GET like Net Framework

In Net Framework, PUT redirects on a 303 do a GET.
Net 5.0 breaks compatibility with this.
See https://github.com/dotnet/runtime/issues/28998
This commit causes redirects of a 303 to do a GET
for POST (existing behaviour), PUT, DELETE, PATCH and OPTIONS

Fix #28998

Commit:
Make 303 redirects do GET like Net Framework

In Net Framework, PUT redirects on a 303 do a GET.
Net 5.0 breaks compatibility with this.
See https://github.com/dotnet/runtime/issues/28998
This commit causes redirects of a 303 to do a GET for all methods except HEAD.

Fix #28998

wfurt commented:
You should add tests covering code changes @TimothyByrd. I agree with @geoffkizer that we should reach consensus on #28998 first.

TimothyByrd commented:
> You should add tests covering code changes @TimothyByrd. I agree with @geoffkizer that we should reach consensus on #28998 first.

I did. They are in src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.AutoRedirect.cs

geoffkizer commented on src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.AutoRedirect.cs:
```diff
@@ -45,6 +45,10 @@ public static IEnumerable<object[]> RedirectStatusCodesOldMethodsNewMethods()
+                yield return new object[] { statusCode, "PUT", statusCode == 303 ? "GET" : "PUT" };
+                yield return new object[] { statusCode, "DELETE", statusCode == 303 ? "GET" : "DELETE" };
+                yield return new object[] { statusCode, "PATCH", statusCode == 303 ? "GET" : "PATCH" };
+                yield return new object[] { statusCode, "OPTIONS", statusCode == 303 ? "GET" : "OPTIONS" };
```
The one thing I'd add here is a completely custom method, e.g. "MYCUSTOMMETHOD" or whatever, and validate it gets changed to GET on 303.

geoffkizer commented:
Overall this looks good, just one comment to add an addition test case above.

Commit:
Add test for MYCUSTOMMETHOD to check 303 redirect

Fix #28998

Commit:
Group test code by behavior

Fix #28998

geoffkizer commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs:
```diff
@@ -142,9 +142,10 @@ private static bool RequestRequiresForceGet(HttpStatusCode statusCode, HttpMetho
                 case HttpStatusCode.MultipleChoices:
                     return requestMethod == HttpMethod.Post;
+                case HttpStatusCode.SeeOther:
+                    return requestMethod != HttpMethod.Head;
```
One more nit: I think this should return false when requestMethod == HttpMethod.Get. 

It probably doesn't make much difference, but it seems better to treat this as not changing the method, as we do with GET in other cases like Moved. If nothing else, it will avoid an unnecessary and confusing trace call.

TimothyByrd replied:
Okay - done.

Commit:
Do not change GET to GET on 303

Fix #28998

geoffkizer commented:
CI failures look unrelated, will merge

geoffkizer commented:
Thanks for your contribution, @TimothyByrd!
